### ISSUE TITLE
Support for multi-platform?

### ISSUE THREAD

**Issue by Jehan** [flatpak/flatpak#560]
It would be interesting to have proper docs for cross-building a software for all platforms supported by flatpak. In my case, I would be interested to build GIMP for at least x86-64 and i386 architectures.

Does Flatpak provide any facilities?

1/ User-side, can we provide a single repository with both a 64bit and a 32bit flatpak (and flatpak on every machine will pull by default the native build without bothering the user with the presence of another package for another arch)?

2/ Builder-side, can I just build with org.gnome.Platform/i386/3.20 and probably -m32 CFLAGS from a x86-64 machine? For this, I assume I would use the --arch=i386 option for the otherwise same build-init command (but inside a different build directory).

In the end, could I export this 32bit build to the same Flatpak repository as the 64bit one to obtain the first point?
Basically some docs would be very useful.

P.S.: I know that I could just test the above, but building the app takes so much time, and I did it so many times in the last weeks (actually as we speak I am building our flatpak again), often in a trial and error approach with undocumented (or not enough) features, that I try to take the easy way for once. Sorry. :-P

**Comment by alexlarsson** [comment]
I'm not sure how to reply to this other than, "yes".
The sdk.gnome.org repos for instance has i386, x86-64, armv7 and aarch64 builds, all in the same repo. And both the i386 and the x86-64 ones are built on the same machine, the only difference is the --arch argument used.

Some modules do need some special handling for i386 on x86-64. see e.g. babl module in https://git.gnome.org/browse/gnome-apps-nightly/tree/org.gnome.Photos.json

**Comment by Jehan** [comment]
> I'm not sure how to reply to this other than, "yes".

Well that's a good answer already. :-)

> Some modules do need some special handling for i386 on x86-64. see e.g. babl module in

Ok. That's a very good example file. I didn't realize that you could set "config-opts" under "arch" (only when set per-module, I guess), and I imagine now any other option which can go under a given module (even maybe additional "sources" for a specific arch, like I could definitely imagine a patch for a bug which only happens with a given architecture, etc.).
Your doc only says about "arch":

>     This is a dictionary defining for each arch a separate build options object that override the main one.

And build options allow c(xx)?flags, env, these kinds of parameters, but not "config-opts" or "sources", etc.

Actually I didn't even realize I could use the same file. I thought I would have to copy the manifest adding the "arch" option with other values. Now that sounds silly, but adding a second arch value (it shows only options for "x86_64") in the flatpak-builder manifest example in the docs (http://flatpak.org/flatpak/flatpak-docs.html#flatpak-builder) would really help unlock this stupid thought. That's by seeing 2 archs in the Photos manifest example that I realized.

These are the kind of stuff that would really help in the documentation. :-)
And in the tutorial-style part too.

**Comment by alexlarsson** [comment]
Oh, the build options unfortunately can't have anything in it, just a few things. It seems like the config-opts part was missing from the docs though. I'll add that.

**Comment by alexlarsson** [comment]
Some discussion on adding more generic "per-arch" stuff in #468
